docs: transform and filter documentation (#2653)
kyma-project/telemetry-manager: docs/user/troubleshooting.md

4. Ensure the OTLP backend service is running and accessible from within the cluster. You can test connectivity from the gateway pod using tools like `curl` (for HTTP) or `grpcurl` (for gRPC).

## Transform or Filter Rules Not Working

### Symptom

You have configured a `transform` or `filter` section in your pipeline, but the data arriving at your backend is not modified, or data you expect to be dropped is still present.

### Cause

This usually happens for one of the following reasons:
* **Incorrect Execution Order**: You're filtering data based on an field's original value, but a transformation rule has already changed it. Transformation rules always run before filter rules.
* **Incorrect Context Path**: Your OTTL expression references a field without the required explicit context path (for example, using `attributes[...]` instead of `resource.attributes[...]`).
* **Condition Never Met**: The condition in your rule is valid (otherwise, you'd see pipeline status: `OTTLSpecInvalid`) but never finds a match in the data. This is often due to a case-sensitive value mismatch or a flawed regular expression.

### Solution

> 1. Review your rules and verify the execution order. For example, if you have a transform rule that renames `resource.attributes["foo"]` to `resource.attributes["bar"]`, your filter rule must check for “bar”, not “foo”.

Reviewer: ```suggestion
1. Review your rules and verify the execution order. For example, if you have a transform rule that renames `resource.attributes["foo"]` to `resource.attributes["bar"]`, your filter rule must check for `bar`. not `foo`.
```

2. **Verify the Execution Order**: Review your rules. If you have a transform that renames resource.attributes["foo"] to resource.attributes["bar"], your filter rule must check for bar, not foo.
3. **Verify Context Paths in Filters**: Ensure your filter conditions use the full, explicit path to the attribute.

>    * **Incorrect**: `attributes["k8s.namespace.name"] == "default"`

Reviewer 2: This incorrect expression should be rejected with a clear error message, so there should be no healthy pipeline. -> It is a different troubleshooting scenario as described

- Reduce emitted data by re-configuring the pipeline (for example, by disabling certain inputs or applying namespace filters).
- Reduce the amount of log data generated by your applications.

## EOF Error When Exporting OTLP Data

### Symptom

The Telemetry gateway logs show errors like: `OTLP export failed: ... rpc error: code = Unavailable desc = error reading from server: EOF`.

### Cause

An "End of File" (EOF) error indicates that the connection to your OTLP backend was closed unexpectedly by the server. This can be caused by several issues:
* **Incorrect Endpoint Address**: The gateway is connecting to the wrong host or port.
* **TLS Configuration Mismatch**: The gateway's TLS settings (`insecure`, `ca`, `cert`, `key`) do not match what the backend server expects. For example, you are connecting without TLS (`insecure: true`) to a server that requires it.
* **Network Policy**: A Kubernetes `NetworkPolicy` is blocking egress traffic from the `telemetry-operator-gateway` pods to the OTLP endpoint.
* **Backend Unavailability**: The backend service is crashing, restarting, or is otherwise unhealthy, causing it to drop active connections.

### Solution

> 1. **Verify the Endpoint**: Double-check the `endpoint` value in your pipeline's `output.otlp` section. Ensure the host and port are correct.

Reviewer 3: ```suggestion
1. Check the `endpoint` value in your pipeline's `output.otlp` section and ensure the host and port are correct.
```

2. **Verify TLS Configuration**:
   * If your backend requires TLS, ensure `insecure` is not set to `true`.
   * If your backend uses a custom Certificate Authority (CA), ensure the `ca` reference in your `tls` configuration is correct.
   * If your backend requires mutual TLS (mTLS), verify that your `cert` and `key` references are correct and the secrets exist.

> 3. **Check for Network Policies**: Look for `NetworkPolicy` resources in the `kyma-system` namespace that could be blocking egress traffic from the gateway pods. You may need to create a policy to explicitly allow this traffic.

Reviewer 3: ```suggestion
3. Look for `NetworkPolicy` resources in the `kyma-system` namespace that could be blocking egress traffic from the gateway pods. You may need to create a policy to explicitly allow this traffic.
```

4. **Check Backend Health**: Ensure the OTLP backend service is running and accessible from within the cluster. You can test connectivity from the gateway pod using tools like `curl` (for HTTP) or `grpcurl` (for gRPC).

## Transform or Filter Rules Not Working

### Symptom

You have configured a `transform` or `filter` section in your pipeline, but the data arriving at your backend is not modified, or data you expect to be dropped is still present.

### Cause

This usually happens for one of the following reasons:

> * **Incorrect Execution Order**: You're filtering based on an attribute's original value, but a transformation has already changed it. Transform rules always run before filter rules.

Reviewer 3: ```suggestion
* **Incorrect Execution Order**: You're filtering based on an attribute's original value, but a transformation rule has already changed it. Transformation rules always run before filter rules.
```

Author: Same here

* **Incorrect Context Path**: Your expression references an attribute without the required explicit context path (for example, using `attributes[...]` instead of `resource.attributes[...]`).

> * **Logic Mismatch**: The condition is syntactically valid, but never evaluates to true (for transforms with conditions) or false (for filters that should drop data) due to value mismatch or regex that doesn't match.

Reviewer 3: ```suggestion
* ** Condition Never Met**: The condition in your rule is valid but never finds a match in the data. This is often due to a case-sensitive value mismatch or a flawed regular expression.
```

> > **Note**: If the OTTL syntax is invalid, the pipeline configuration is not generated. The corresponding error message appears in the pipeline status under the `ConfigurationGenerated` condition with reason `OTTLSpecInvalid`.

Author: Initially, I thought that syntax errors would be ignored, but we handle them in the reconciliation loop, meaning it will be reflected in the status of the pipeline, under the ConfigurationGenerated condition.

I just wanted to highlight that behavior here so that it is clear for the user that this issue is covered. It is not really a cause of the "Transform or Filter Rules Not Working", since the whole configuration is not applied at all in this case, so you know for sure that they are not syntactically correct i.e. will not be working.

> > **Note**: If the OTTL syntax is invalid, the pipeline configuration is not generated. The corresponding error message appears in the pipeline status under the `ConfigurationGenerated` condition with reason `OTTLSpecInvalid`.

Author: But it can as well just be another bullet point. Lmk what you think

### Solution

> 1. **Verify the Execution Order**: If a transform renames `resource.attributes["foo"]` to `resource.attributes["bar"]`, subsequent filter rules must reference `bar`.

Reviewer 3: ```suggestion
1. Review your rules and verify the execution order. For example, if you have a transform rule that renames `resource.attributes["foo"]` to `resource.attributes["bar"]`, your filter rule must check for “bar”, not “foo”.
```

4. Ensure the OTLP backend service is running and accessible from within the cluster. You can test connectivity from the gateway pod using tools like `curl` (for HTTP) or `grpcurl` (for gRPC).

## Transform or Filter Rules Not Working

### Symptom

You have configured a `transform` or `filter` section in your pipeline, but the data arriving at your backend is not modified, or data you expect to be dropped is still present.

### Cause

This usually happens for one of the following reasons:
* **Incorrect Execution Order**: You're filtering based on an attribute's original value, but a transformation has already changed it. Transform rules always run before filter rules.
* **Incorrect Context Path**: Your OTTL expression references an attribute without the required explicit context path (for example, using `attributes[...]` instead of `resource.attributes[...]`).
* **Condition Never Met**: The condition in your rule is valid but never finds a match in the data. This is often due to a case-sensitive value mismatch or a flawed regular expression.

> [!NOTE]

> > If the OTTL syntax is invalid, the pipeline configuration is not generated. The corresponding error message appears in the pipeline status under the `ConfigurationGenerated` condition with reason `OTTLSpecInvalid`.

Reviewer 3: If the OTTL syntax is invalid, is then the symptom that the transform/filter rule doesn't work? (the headline of this section)
Or would invalid OTTL syntax lead to another symptom? (such as: no pipeline exists (?), pipeline status "OTTLSpecInvalid").

=> Does this error truly belong under this headline?

Reviewer 3: Decision: [Include it in the sentence above](https://github.com/kyma-project/telemetry-manager/pull/2653#discussion_r2499974376) + drop this note.

- Reduce emitted data by re-configuring the pipeline (for example, by disabling certain inputs or applying namespace filters).
- Reduce the amount of log data generated by your applications.

## Transform or Filter Rule Has No Effect

### Symptom

You have configured a `transform` or `filter` section in your pipeline, but the data arriving at your backend is not modified, or data you expect to be dropped is still present.

### Cause

This usually happens for one of the following reasons:
* **Incorrect Execution Order**: You're filtering data based on a field's original value, but a transformation rule has already changed it. Transformation rules always run before filter rules.

> * **Condition Never Met**: The condition in your rule is valid (otherwise, the pipeline condition `ConfigurationGenerated` would have a `False` status with the reason `OTTLSpecInvalid`), but never finds a match in the data. This is often due to a case-sensitive value mismatch or a flawed regular expression.

Reviewer 3: ```suggestion
* **Condition Never Met**: The condition in your rule is valid (otherwise, you'd see pipeline condition `ConfigurationGenerated`:`False` with the reason `OTTLSpecInvalid`), but never finds a match in the data. This is often due to a case-sensitive value mismatch or a flawed regular expression.
```

### Solution

1. Review your rules and verify the execution order. For example, if you have a transform rule that renames `resource.attributes["foo"]` to `resource.attributes["bar"]`, your filter rule must check for `bar`, not `foo`.
2. **Inspect Logic / Regex**: Test your regex separately. Simplify complex conditions to a single comparison and re-apply.
3. To test your rules, temporarily remove all but one rule to confirm it works as expected. Then, add your other rules incrementally and isolate the rule that is causing the issue.

> ## Vague EOF Error Message When Using OTTL Transform or Filter

Reviewer 3: ```suggestion
## Generic EOF Error Message When Using OTTL Transform or Filter
```

### Symptom

> Pipeline configuration fails with vague error messages mentioning "unexpected token `<EOF>`" or EOF (End of File) parsing errors, and the pipeline status shows `ConfigurationGenerated` condition with status `False` and reason `OTTLSpecInvalid`.

Reviewer 3: ```suggestion
- The pipeline configuration fails with vague error messages mentioning "unexpected token `<EOF>`" or EOF (End of File) parsing errors.
- In the pipeline status, you see the condition `ConfigurationGenerated` with status `False` and reason `OTTLSpecInvalid`.
```

### Cause

This error occurs due to inadequately processed syntax errors in your OTTL (OpenTelemetry Transformation Language) transform and/or filter rules. Under normal circumstances, syntax validation would provide a descriptive error message explicitly identifying the issue in your configuration. However, this represents an edge case where the validation process fails to generate meaningful diagnostic information, resulting in the generic "unexpected token `<EOF>`" message. Common underlying causes include:

- **Missing closing parentheses** or **brackets** in function calls or conditions
- **Incorrect function names** or **typos** in OTTL function syntax
- **Malformed string literals** or **regex patterns**
- **Invalid operator usage** or **missing operators** in conditional expressions

### Solution

> 1. **Double-check the syntax** of your transform and filter rules: Verify all parentheses and brackets are properly closed, ensure function names are spelled correctly (for example, `IsMatch()` not `sIsMatch()`), and validate that you're using the correct OTTL function syntax as documented in the [OTTL function reference](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md).

Reviewer 3: ```suggestion
1. Review the syntax of your transform and filter rules: Verify all parentheses and brackets are closed, ensure function names are spelled correctly (for example, `IsMatch()` not `sIsMatch()`), and validate that you're using the correct OTTL function syntax (see [OTTL Functions](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md)).
```
